skills: add basic methods for skill handing

zyga opened the pull request:
This patch adds five related functions:

Repository.AddSkill() adds a new skill to the repository.
Repository.RemoveSkill() removes a skill with by snap and skill name.
Repository.Skill() looks up a skill by snap and skill name.
Repository.AllSkills() returns a list of skills of a given type.
Repository.AllSkills() can also be used to look for any type, returning
all the skills in general.
Repository.Skills() returns a list of skills offered by a snap.
    
Internally there's also a helper Repository.unlockedSkill() that finds a
skill without holding a lock and a sort interface to keep the list of
skills ordered by snap and skill name.

zyga commented on skills/repo.go:
```diff
@@ -19,10 +19,201 @@
+	defer r.m.Unlock()
+
+	// TODO: Ensure that the skill is not used anywhere
+	for i, skill := range r.skills {
```
NOTE: this is fixed to use binary search in the grant-revoke branch (pull #374 )

niemeyer replied:
That can actually be a trivial map operation, as suggested. Let's please do it in this branch.

zyga replied:
Done

niemeyer commented:
Will wait until we sort out the pre-reqs.

niemeyer commented on skills/repo.go:
```diff
@@ -19,10 +19,200 @@
+	// ErrSkillNotFound is reported when skill cannot be looked up.
+	ErrSkillNotFound = errors.New("skill not found")
+	// ErrDuplicateSkill is reported when skill with duplicate name is being added to a repository.
+	ErrDuplicateSkill = errors.New("duplicate skill name")
```
Drop all of these please. Replace them by inline fmt.Errorf with proper information reported.

zyga replied:
Done

niemeyer commented on skills/repo.go:
```diff
@@ -19,10 +19,200 @@
+	// Protects the internals from concurrent access.
+	m      sync.Mutex
+	types  []Type
+	skills []*Skill
```
```
skills map[string][string]*Skill
slots map[string][string]*Slot

skills[snapName][skillName]
slots[snapName][slotName]
```

zyga replied:
Done

niemeyer commented on skills/repo.go:
```diff
@@ -19,10 +19,200 @@
+// AddSkill adds a skill to the repository.
+// Skill names must be valid snap names, as defined by ValidateName.
+// Skill name must be unique within a particular snap.
+func (r *Repository) AddSkill(snapName, skillName, typeName, label string, attrs map[string]interface{}) error {
```
What happens when there's a new field?

zyga replied:
I've changed this to AddSkill(skill *Skill) error

niemeyer commented on skills/repo.go:
```diff
@@ -19,10 +19,200 @@
+	defer r.m.Unlock()
+
+	// Reject skill with invalid names
+	if err := ValidateName(snapName); err != nil {
```
That's a skill name validator, not a snap name.. do we have one for snaps?

zyga replied:
That's a good point. I'm not sure we do (grepping quickly I didn't find one). Since we don't really look at snap names much, perhaps we don't really need to validate it?

niemeyer commented on skills/repo.go:
```diff
@@ -19,10 +19,200 @@
+	if err := ValidateName(skillName); err != nil {
+		return err
+	}
+	// TODO: ensure that given snap really exists
```
It's not its job to control system state. It's been explicitly told to add that information as valid.

zyga replied:
Removed

niemeyer commented on skills/repo.go:
```diff
@@ -19,10 +19,200 @@
+	return nil
+}
+
+func (r *Repository) unlockedSkillIndex(snapName, skillName string) (int, bool) {
```
That's r.skills[snapName][skillName]

niemeyer commented on skills/repo.go:
```diff
@@ -19,10 +19,200 @@
+	return i, false
+}
+
+func (r *Repository) unlockedSkill(snapName, skillName string) *Skill {
```
That's r.skills[snapName][skillName]

zyga replied:
Done

zyga commented:
I've addressed most of the comments. I need to do a pass on documentation to remove references to Err* and there's one thing I'm not sure about wrt snap name validation.

zyga commented:
I've made the snap name validation separate. Please have a 2nd look now.

mvo5 commented on skills/repo.go:
```diff
@@ -29,12 +32,15 @@ type Repository struct {
 	// Protects the internals from concurrent access.
 	m     sync.Mutex
 	types map[string]Type
+	// Indexed by [snapName][skillName]
```
Nice, thanks for the comment

mvo5 commented:
LGTM, thanks for adding the snap validate.go. Once that lands I will use it to the snap.yaml code.

niemeyer commented on skills/repo.go:
```diff
@@ -61,3 +67,102 @@ func (r *Repository) AddType(t Type) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return r.skills[snapName][skillName]
```
This will panic if the snapName isn't found.

niemeyer replied:
Oh, no it won't! I just learned that indexing a nil map is okay.

niemeyer commented on skills/repo.go:
```diff
@@ -61,3 +67,102 @@ func (r *Repository) AddType(t Type) error {
+	defer r.m.Unlock()
+
+	// Reject snaps with invalid names
+	if err := snap.ValidateName(skill.Snap); err != nil {
```
Thanks for that.

niemeyer commented on skills/repo.go:
```diff
@@ -61,3 +67,102 @@ func (r *Repository) AddType(t Type) error {
+		return err
+	}
+	if _, ok := r.skills[skill.Snap][skill.Name]; ok {
+		return fmt.Errorf("cannot add skill, skill name %q is in use", skill.Name)
```
"cannot add skill, snap %q already has a skill %q"

niemeyer commented on skills/repo.go:
```diff
@@ -61,3 +67,102 @@ func (r *Repository) AddType(t Type) error {
+
+	// TODO: Ensure that the skill is not used anywhere
+	if _, ok := r.skills[snapName][skillName]; !ok {
+		return fmt.Errorf("cannot remove skill %q, no such skill", skillName)
```
s/skill %q/skill %q from snap %q"

niemeyer commented on skills/repo.go:
```diff
@@ -61,3 +67,102 @@ func (r *Repository) AddType(t Type) error {
+	if _, ok := r.skills[snapName][skillName]; !ok {
+		return fmt.Errorf("cannot remove skill %q, no such skill", skillName)
+	}
+	delete(r.skills[snapName], skillName)
```
`if len(r.skills[snapName]) == 0 { delete(r.skills, snapName) }`

So it doesn't grow unbounded over time..

zyga replied:
Ah, indeed. Thanks for spotting this.

niemeyer commented on skills/repo.go:
```diff
@@ -61,3 +67,102 @@ func (r *Repository) AddType(t Type) error {
+			}
+		}
+	}
+	sort.Sort(bySkillSnapAndName(result))
```
The sorting here will feel slightly awkward, potentially. We can run with that for now and watch the use cases.

zyga replied:
I agree, I prefer to sort by default for unit tests (far less annoying) but perhaps there's no real need for this.

niemeyer commented:
LGTM, thank you! Just a few trivials above.
